biomage/stage/stage.py

```diff
@@ -224,12 +263,307 @@ def create_manifest(templates, token):
+
+            # Switch according to chosen action
+            if page_action == "next":
+                current_page += 1
+            elif page_action == "previous":
+                current_page -= 1
+            elif page_action == "done":
+                break
+
+    return chosen_experiments
+
+
+def check_file_etag(target, etag):
+    """
+    Copy s3 files in a bucket under a prefix
+    """
+    same_etag = False
+
+    try:
+        s3 = boto3.client("s3")
+        s3.head_object(Bucket=target["Bucket"], Key=target["Key"], IfMatch=etag)
+        same_etag = True
+    except ClientError as err:
+        if err.response['ResponseMetadata']['HTTPStatusCode'] not in [412, 404]:
+            raise
+
+    return same_etag
+
+
+def copy_s3_files(sandbox_id, prefix, source_bucket, target_bucket):
+    """
+    Copy s3 files in a bucket under a prefix
+    """
+    s3 = boto3.client("s3")
+    exp_files = s3.list_objects_v2(Bucket=source_bucket, Prefix=prefix)
+
+    for obj in exp_files.get("Contents"):
+
+        experiment_id = obj["Key"].split("/")[0]
+
+        source = {"Bucket": source_bucket, "Key": obj["Key"]}
+
+        target = {
+            "Bucket": target_bucket,
+            "Key": obj["Key"].replace(experiment_id, f"{sandbox_id}-{experiment_id}"),
+        }
+
+        if not check_file_etag(target, obj["ETag"]):
+            try:
+                click.echo(
+                    f"Copying from {source['Bucket']}/{source['Key']} to "
+                    f"{target['Bucket']}/{target['Key']}"
+                )
+
+                s3.copy_object(
+                    CopySource=source,
+                    Bucket=target["Bucket"],
+                    Key=target["Key"],
+                )
+            except Exception as e:
+                click.echo(
+                    f"failed to copy object {source['Bucket']}/{source['Key']} \
+                    with exception: \n {e}"
+                )
+
+
+def copy_dynamodb_records(sandbox_id, staging_experiments, source_table, target_table):
+    """
+    Copy dynamodBD records for an experiment id
+    """
+    dynamodb = boto3.client("dynamodb")
+    for experiment_id in staging_experiments:
+        items = dynamodb.query(
+            TableName=source_table,
+            KeyConditionExpression="experimentId = :experiment_id",
+            ExpressionAttributeValues={":experiment_id": {"S": experiment_id}},
+        ).get("Items")
+
+        items_to_insert = {}
+        items_to_insert[target_table] = [
+            {
+                "PutRequest": {
+                    "Item": {
+                        **item,
+                        "experimentId": {
+                            "S": f"{sandbox_id}-{item['experimentId']['S']}",
+                        },
+                    }
+                }
+            }
+            for item in items
+        ]
+
+        try:
+            dynamodb.batch_write_item(RequestItems=items_to_insert)
+        except Exception as e:
```

Comment:
if you had any specific exception during dev, add it here instead of `Exception` generic one. Otherwise just leave it

Reply:
No, I haven't found any specific exception for this one... It's just a catch-all in case of something up, as we're interacting with an external service